vllm/model_executor/layers/quantization/fp8.py

```diff
@@ -0,0 +1,148 @@
+        # https://github.com/pytorch/pytorch/pull/118881/files
+        return 89
+
+    @classmethod
+    def get_config_filenames(cls) -> List[str]:
+        return []
+
+    @classmethod
+    def from_config(cls, config: Dict[str, Any]) -> "FP8Config":
+        return cls()
+
+    def get_linear_method(self) -> "Fp8LinearMethod":
+        return Fp8LinearMethod(self)
+
+    def get_scaled_act_names(self) -> List[str]:
+        return []
+
+
+class Fp8LinearState(Enum):
+
+    UNINITIALIZED = enum.auto()
+    READY = enum.auto()
+
+
+class Fp8LinearMethod(LinearMethodBase):
+    """Linear method for FP8.
+    We now support two types of model checkpoints:
+    1. Common FP16/BF16 model checkpoints. In this case, the weight
+       scaling factor will be initialized during the first forward pass.
+    2. FP8 model checkpoints. In this case, the weight scaling factor
+       will be loaded from the checkpoint with parameter name "w_scale",
+       and the weight should already in FP8 and has been transposed.
+
+    Note that we currently only support per-tensor quantization.
+       
+    Args:
+        quant_config: The quantization config.
+    """
+
+    def __init__(self, quant_config: FP8Config):
+        self.quant_config = quant_config
+
+    def create_weights(
+        self,
+        layer: torch.nn.Module,
+        input_size_per_partition: int,
+        output_size_per_partition: int,
+        input_size: int,
+        output_size: int,
+        params_dtype: torch.dtype,
+        **extra_weight_attrs,
+    ):
+        weight = Parameter(torch.empty(output_size_per_partition,
+                                       input_size_per_partition,
+                                       dtype=params_dtype),
+                           requires_grad=False)
+        set_weight_attrs(weight, {"input_dim": 1, "output_dim": 0})
+        layer.register_parameter("weight", weight)
+        set_weight_attrs(weight, extra_weight_attrs)
+
+        # Will be loaded from FP8 checkpoints, or initialized for
+        # FP16/BF16 checkpoints during the first forward pass.
+        w_scale = Parameter(
+            torch.empty(1, dtype=torch.float32),
+            requires_grad=False,
+        )
+        layer.register_parameter("w_scale", w_scale)
+        set_weight_attrs(w_scale, extra_weight_attrs)
+        layer.fp8_linear_state = Fp8LinearState.UNINITIALIZED
+
+    def apply_weights(self,
+                      layer: torch.nn.Module,
+                      x: torch.Tensor,
+                      bias: Optional[torch.Tensor] = None) -> torch.Tensor:
+
+        if layer.fp8_linear_state == Fp8LinearState.UNINITIALIZED:
+            # Per-tensor scaling on the fly for FP16/BF16 weights
+            # during the first forward pass.
+            if layer.weight.dtype != torch.float8_e4m3fn:
+                qweight, weight_scale = per_tensor_quantize(layer.weight.data)
+                layer.weight.data = qweight.t()
```

Comment:
Can you say why the tensor is being transposed here? Does it have something to do with `_scaled_mm` or is it also done for normal linear layers?

Reply:
This is because _scaled_mm requires column major layout.

Reply:
(it wouldn't hurt to add a comment in the code that clarifies this)

Reply:
I did some tests yesterday and it seems ok. When param.data is overwritten, the original tensor has no reference anymore and will be freed (torch.cuda.empty_cache may be required to reduce fragmentation).

I guess my impression comes from training, and our weights are frozen all the time in inference...